### PR TITLE
add event_expansion for v1beta1 Events

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

/assign @wojtek-t 
/sig scalability

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by yastij** [comment]
/priority important-soon

@wojtek-t - can you tag this for 1.15 ?

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/generated_expansion.go [hunk 1]
```diff
@@ -17,5 +17,3 @@ limitations under the License.
 // Code generated by client-gen. DO NOT EDIT.
 
 package v1beta1
-
-type EventExpansion interface{}
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/generated_expansion.go]: This file can't be edited manually.

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+		NamespaceIfScoped(event.Namespace, len(event.Namespace) > 0).
+		Resource("events").
+		Name(event.Name).
+		Body(data).
+		Do().
+		Into(result)
+	return result, err
+}
+
+type EventSinkImpl struct {
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: Hmm - I'm wondering why we need that.
I'm looking into core/v1 Events and I'm seeing this TODO:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go#L149

With the new API we have full control - we can migrate events one by one and start only with those that are using client-go.
So why do we need that at all?

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+package v1beta1
+
+import (
+	"fmt"
+
+	"k8s.io/api/events/v1beta1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// TODO: Add querying functions to the event expansion
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: nit: move todo below the comment for the interface.

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// The EventExpansion interface allows manually adding extra methods to the EventInterface.
+type EventExpansion interface {
+	// CreateWithEventNamespace is the same as a Create, except that it sends the request to the event.Namespace.
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: nit: can you please split those lines to make them a bit shorter.

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+		NamespaceIfScoped(event.Namespace, len(event.Namespace) > 0).
+		Resource("events").
+		Name(event.Name).
+		Body(data).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// TODO: this is easier to implement this as a sink given that the passed
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: TBH, I can't parse this comment, I'm not really following what this means :)
[I buy the argument with push to clientset being transparent though]

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+		Name(event.Name).
+		Body(data).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// EventSinkImpl wraps EventInterface to implement EventSink
+// TODO: this makes it easier for testing purpose and masks the logic of performing API calls
+// note that rollbacking to raw clientset should also be transparent
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: s/note/Note/

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+		Resource("events").
+		Name(event.Name).
+		Body(data).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// EventSinkImpl wraps EventInterface to implement EventSink
+// TODO: this makes it easier for testing purpose and masks the logic of performing API calls
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: "." at the end of the sentence.

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+		NamespaceIfScoped(event.Namespace, len(event.Namespace) > 0).
+		Resource("events").
+		Name(event.Name).
+		Body(data).
+		Do().
+		Into(result)
+	return result, err
+}
+
+// EventSinkImpl wraps EventInterface to implement EventSink
```
Comment (wojtek-t) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: Why this lives here?

I think this should be move to where recorder is implemented.

**Comment by wojtek-t** [comment]
@yastij - please fix gofmt and then lgtm

**Comment by yastij** [comment]
@wojtek-t - fixed

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+// The EventExpansion interface allows manually adding extra methods to the EventInterface.
+// TODO: Add querying functions to the event expansion
+type EventExpansion interface {
+	// CreateWithEventNamespace is the same as a Create
+	// except that it sends the request to the event.Namespace.
+	CreateWithEventNamespace(event *v1beta1.Event) (*v1beta1.Event, error)
+	// UpdateWithEventNamespace is the same as a Update
+	// except that it sends the request to the event.Namespace.
+	UpdateWithEventNamespace(event *v1beta1.Event) (*v1beta1.Event, error)
+	// PatchWithEventNamespace is the same as a Update
```
Comment (tedyu) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: a Update -> an Update

File: staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+	PatchWithEventNamespace(event *v1beta1.Event, data []byte) (*v1beta1.Event, error)
+}
+
+// CreateWithEventNamespace makes a new event.
+// Returns the copy of the event the server returns, or an error.
+// The namespace to create the event within is deduced from the event.
+// it must either match this event client's namespace, or this event client must
+// have been created with the "" namespace.
+func (e *events) CreateWithEventNamespace(event *v1beta1.Event) (*v1beta1.Event, error) {
+	if e.ns != "" && event.Namespace != e.ns {
```
Comment (tedyu) [inline staging/src/k8s.io/client-go/kubernetes/typed/events/v1beta1/event_expansion.go]: what if len(event.Namespace) == 0 ?

**Pull request merged**